File: bidirectional_astar.py
# -*- coding: utf-8 -*-
import heapq
import time

class BidirectionalAStar:
    def __init__(self, grid):
        self.grid = grid
        self.height = len(grid)
        self.width = len(grid[0])
        self.movements = [(0, 1), (1, 0), (0, -1), (-1, 0),
                         (1, 1), (1, -1), (-1, 1), (-1, -1)]
        self.nodes_explored = 0
        self.execution_time = 0
        self.path_length = 0

    class Node:
        def __init__(self, x, y, parent=None, is_forward=True):
            self.x = x
            self.y = y
            self.parent = parent
            self.g = 0
            self.h = 0
            self.f = 0
            self.is_forward = is_forward  # 标记搜索方向

        def __lt__(self, other):
            return self.f < other.f

    def heuristic(self, node, goal):
        dx = abs(node.x - goal.x)
        dy = abs(node.y - goal.y)
        return 10 * (dx + dy) + (14 - 2 * 10) * min(dx, dy)

    def find_path(self, start, end):
        start_time = time.time()
        self.nodes_explored = 0
        
        # 初始化正向和反向搜索
        forward_open = []
        backward_open = []
        start_node = self.Node(*start, is_forward=True)
        end_node = self.Node(*end, is_forward=False)
        heapq.heappush(forward_open, start_node)
        heapq.heappush(backward_open, end_node)
        
        forward_closed = dict()
        backward_closed = dict()

        while forward_open and backward_open:
            # 处理正向搜索
            current_forward = heapq.heappop(forward_open)
            self.nodes_explored += 1

            # 检查相遇条件
            if (current_forward.x, current_forward.y) in backward_closed:
                meeting_node = backward_closed[(current_forward.x, current_forward.y)]
                path = self._merge_paths(current_forward, meeting_node)
                self._finalize_stats(start_time, path)
                return path

            if (current_forward.x, current_forward.y) not in forward_closed:
                forward_closed[(current_forward.x, current_forward.y)] = current_forward
                self._expand_node(current_forward, forward_open, forward_closed, backward_closed, is_forward=True, goal=end_node)

            # 处理反向搜索
            current_backward = heapq.heappop(backward_open)
            self.nodes_explored += 1

            # 检查相遇条件
            if (current_backward.x, current_backward.y) in forward_closed:
                meeting_node = forward_closed[(current_backward.x, current_backward.y)]
                path = self._merge_paths(meeting_node, current_backward)
                self._finalize_stats(start_time, path)
                return path

            if (current_backward.x, current_backward.y) not in backward_closed:
                backward_closed[(current_backward.x, current_backward.y)] = current_backward
                self._expand_node(current_backward, backward_open, backward_closed, forward_closed, is_forward=False, goal=start_node)

        self.execution_time = time.time() - start_time
        return None

    def _expand_node(self, current, open_list, closed_dict, other_closed, is_forward, goal):
        for dx, dy in self.movements:
            nx = current.x + dx
            ny = current.y + dy
            if 0 <= nx < self.height and 0 <= ny < self.width:
                # 对角线移动障碍检查
                if dx != 0 and dy != 0:
                    if self.grid[current.x + dx][current.y] == 1 or \
                       self.grid[current.x][current.y + dy] == 1:
                        continue
                
                if self.grid[nx][ny] == 1 or (nx, ny) in closed_dict:
                    continue

                # 计算移动成本
                move_cost = 14 if dx != 0 and dy != 0 else 10
                new_node = self.Node(nx, ny, current, is_forward)
                new_node.g = current.g + move_cost
                new_node.h = self.heuristic(new_node, goal)
                new_node.f = new_node.g + new_node.h

                heapq.heappush(open_list, new_node)

    def _merge_paths(self, forward_node, backward_node):
        # 合并正向和反向路径
        forward_path = []
        current = forward_node
        while current:
            forward_path.append((current.x, current.y))
            current = current.parent

        backward_path = []
        current = backward_node
        while current:
            backward_path.append((current.x, current.y))
            current = current.parent

        # 合并并去除重复点
        return forward_path[::-1] + backward_path[1:]

    def _finalize_stats(self, start_time, path):
        self.execution_time = time.time() - start_time
        if path:
            self.path_length = len(path) 

File: BidirectionalAStar.js
class Node {
    constructor(x, y, parent = null, isForward = true) {
        this.x = x;
        this.y = y;
        this.parent = parent;
        this.g = 0;
        this.h = 0;
        this.f = 0;
        this.isForward = isForward; // Mark search direction
    }

    compareTo(other) {
        return this.f - other.f;
    }
}

class BidirectionalAStar {
    constructor(grid) {
        this.grid = grid;
        this.height = grid.length;
        this.width = grid[0].length;
        this.movements = [[0, 1], [1, 0], [0, -1], [-1, 0], [1, 1], [1, -1], [-1, 1], [-1, -1]];
        this.nodesExplored = 0;
        this.executionTime = 0;
        this.pathLength = 0;
    }

    heuristic(node, goal) {
        const dx = Math.abs(node.x - goal.x);
        const dy = Math.abs(node.y - goal.y);
        return 10 * (dx + dy) + (14 - 20) * Math.min(dx, dy);
    }

    findPath(start, end) {
        const startTime = performance.now();
        this.nodesExplored = 0;

        const forwardOpen = [];
        const backwardOpen = [];
        const startNode = new Node(...start, true);
        const endNode = new Node(...end, false);
        forwardOpen.push(startNode);
        backwardOpen.push(endNode);
        
        const forwardClosed = new Map();
        const backwardClosed = new Map();

        while (forwardOpen.length > 0 && backwardOpen.length > 0) {
            const currentForward = forwardOpen.sort((a, b) => a.compareTo(b)).shift();
            this.nodesExplored += 1;

            if (backwardClosed.has(`${currentForward.x},${currentForward.y}`)) {
                const meetingNode = backwardClosed.get(`${currentForward.x},${currentForward.y}`);
                const path = this._mergePaths(currentForward, meetingNode);
                this.executionTime = performance.now() - startTime;
                return path;
            }

            if (!forwardClosed.has(`${currentForward.x},${currentForward.y}`)) {
                forwardClosed.set(`${currentForward.x},${currentForward.y}`, currentForward);
                this._expandNode(currentForward, forwardOpen, forwardClosed, backwardClosed, true, endNode);
            }

            const currentBackward = backwardOpen.sort((a, b) => a.compareTo(b)).shift();
            this.nodesExplored += 1;

            if (forwardClosed.has(`${currentBackward.x},${currentBackward.y}`)) {
                const meetingNode = forwardClosed.get(`${currentBackward.x},${currentBackward.y}`);
                const path = this._mergePaths(meetingNode, currentBackward);
                this.executionTime = performance.now() - startTime;
                return path;
            }

            if (!backwardClosed.has(`${currentBackward.x},${currentBackward.y}`)) {
                backwardClosed.set(`${currentBackward.x},${currentBackward.y}`, currentBackward);
                this._expandNode(currentBackward, backwardOpen, backwardClosed, forwardClosed, false, startNode);
            }
        }

        this.executionTime = performance.now() - startTime;
        return null;
    }

    _expandNode(current, openList, closedDict, otherClosed, isForward, goal) {
        for (const [dx, dy] of this.movements) {
            const nx = current.x + dx;
            const ny = current.y + dy;
            if (nx >= 0 && nx < this.height && ny >= 0 && ny < this.width) {
                if ((dx !== 0 && dy !== 0) && (this.grid[current.x + dx][current.y] === 1 || this.grid[current.x][current.y + dy] === 1)) 
                    continue;

                if (this.grid[nx][ny] === 1 || otherClosed.has(`${nx},${ny}`)) continue;

                const moveCost = (dx !== 0 && dy !== 0) ? 14 : 10;
                const newNode = new Node(nx, ny, current, isForward);
                newNode.g = current.g + moveCost;
                newNode.h = this.heuristic(newNode, goal);
                newNode.f = newNode.g + newNode.h;

                openList.push(newNode);
            }
        }
    }

    _mergePaths(forwardNode, backwardNode) {
        const forwardPath = [];
        let current = forwardNode;
        while (current) {
            forwardPath.push([current.x, current.y]);
            current = current.parent;
        }

        const backwardPath = [];
        current = backwardNode;
        while (current) {
            backwardPath.push([current.x, current.y]);
            current = current.parent;
        }

        return forwardPath.reverse().concat(backwardPath.slice(1));
    }
}

export { BidirectionalAStar, Node };